ir/src/constraints/list_comprehension.rs

```diff
@@ -0,0 +1,328 @@
+use super::{
+    graph::{CURRENT_ROW, NEXT_ROW},
+    IdentifierType, SemanticError, SymbolTable,
+};
+use air_script_core::{
+    Expression, Identifier, Iterable, ListComprehension, ListFoldingType, NamedTraceAccess,
+    VariableType, VectorAccess,
+};
+
+/// Parses a list comprehension and returns the corresponding vector of expressions.
+///
+/// Returns an error if there is an error while parsing any of the expressions in the expanded
+/// vector from the list comprehension.
+pub fn parse_lc(
+    lc: &ListComprehension,
+    symbol_table: &SymbolTable,
+) -> Result<Vec<Expression>, SemanticError> {
+    let lc_length = lc_length(lc, symbol_table)?;
+    let vector = (0..lc_length)
+        .map(|i| parse_lc_expr(lc.expression(), lc, symbol_table, i))
+        .collect::<Result<Vec<_>, _>>()?;
+    Ok(vector)
+}
+
+/// Parses a list comprehension expression and creates an expression based on the index of the
+/// expression in the list comprehension.
+///
+/// Returns an error if there is an error while parsing the sub-expression.
+fn parse_lc_expr(
+    expression: &Expression,
+    lc: &ListComprehension,
```

Comment:
It seems to me that we are passing `ListComprehension` around to get access to its context and then use loops to figure out which member corresponds to which iterable. If so, a better solution could be to build a map from identifier to iterable upfront (maybe put this map into something a struct - something like `IterableContext`), and then pass this map around instead `ListComprehension`.

This way, instead of doing something like:
```Rust
for (member, iterable) in lc.context() {
    if ident == member {
        match iterable {
```
You'd be able to do something like:
```Rust
let iterable = context.get_iterable(ident)?;
match iterable {
```

You may be able to take it even a step further and put `IdentifierType` for each member into the context. This way, you won't need to pass `symbol_table` around either.

Reply:
Thanks, done. I moved it to just a `BTreeMap<Identifier, Iterable>`. I didn't remove symbol table as it is required to build context for list folding inside list comprehension.